srk/test/test_nestedLoops.ml

```diff
@@ -0,0 +1,577 @@
+    )
+    (mk_true srk)
+    lambdas
+
+let rec build_subtracted_lambdas lambdas1 lambdas2 =
+  match lambdas1, lambdas2 with
+  | h1 :: t1, h2 :: t2 -> 
+    mk_sub srk h1 h2 :: build_subtracted_lambdas t1 t2
+  | [], [] -> []
+  | _, _ -> assert false
+
+let get_constants_vec cs constraints =
+  BatList.fold_left
+    (fun (existing_rows, i) eq -> 
+        match eq with
+        | `LeqZero t -> 
+          begin
+            let row = Vec.coeff CoordinateSystem.const_id t in
+            let neg = QQ.negate row in
+            let m =  Vec.add_term neg i existing_rows in
+            (m, i+1)
+          end
+        | `LtZero t -> Log.error "there should not be strict inequalities"; assert false
+        | `EqZero t -> 
+          begin
+            let row = Vec.coeff CoordinateSystem.const_id t in
+            let neg = QQ.negate row in
+            let m =  Vec.add_term neg i existing_rows in
+            let row2 = Vec.coeff CoordinateSystem.const_id t in
+            let m =  Vec.add_term row2 (i+1) m in
+            (m, i+2)
+          end
+    )
+    (Vec.zero, 0)
+    constraints
+  
+let print_formula_to_stdout descrip f =
+  Format.fprintf Format.std_formatter descrip;
+  let s2 = (Printf.sprintf "\n%s\n" (Formula.show srk f)) in
+  Format.fprintf Format.std_formatter "%s" s2;
+  ()
+
+let print_ranking_func cs xp_list rf_terms_list rf_delta0 rf_delta interp =
+  (* let len = List.length xp_list in *)
+  let rf = List.map
+      (fun term -> Interpretation.evaluate_term interp term)
+      rf_terms_list
+  in
+  Format.fprintf Format.std_formatter "\nrf has %d terms.\n\n" (List.length rf);
+  let d0 = List.map
+    (fun term -> Interpretation.evaluate_term interp term)
+    rf_delta0
+  in
+  let d = List.map
+    (fun term -> Interpretation.evaluate_term interp term)
+    rf_delta
+  in
+  Format.fprintf Format.std_formatter "\n\n********* Ranking function ************\n";
+  Format.fprintf Format.std_formatter "f = ";
+  let n_cols = List.length rf in
+  for i = 0 to n_cols-1 do
+    let real_var = List.nth xp_list i in
+    (* let xp = List.nth xp_list i in *)
+    pp_symbol srk Format.std_formatter real_var;
+    Format.fprintf Format.std_formatter " * ";
+    let coe = List.nth rf i in
+    QQ.pp Format.std_formatter coe;
+    if i<n_cols-1 then
+    Format.fprintf Format.std_formatter " + ";
+  done;
+  Format.fprintf Format.std_formatter "\nIt has to decrese by this amount in each iter: ";
+  QQ.pp Format.std_formatter (QQ.sub QQ.zero (List.nth d 0));
+  logf "\nIt has to decrese by this amount in each iter: %a" QQ.pp (QQ.sub QQ.zero (List.nth d 0)); 
+
+  Format.fprintf Format.std_formatter "\nAnd its lower bound is:\n";
+  let t = QQ.negate (List.nth d0 0) in
+  (* let t = QQ.sub (List.nth d 0) (List.nth d0 0) in *) 
+  QQ.pp Format.std_formatter t;
+  Format.fprintf Format.std_formatter "\n\n";
+
+  Format.fprintf Format.std_formatter "\n\n***************************\n\n";
+  ()
+
+let prove_termination loop =
```

Comment:
After you rewrite your prove_termination procedure to return a result rather than printing to the screen, you can delete all of the duplicated code in test_nestedLoops